### ISSUE TITLE
Again fields in diagnostics.F90

### ISSUE THREAD

**Issue by DanRRRR** [Warwick-Plasma/epoch#482]
I just posted this at the end of Closed Issue and not sure it is visible, so i re-post it here

I have even more strange related to the fields problem exactly in this subroutine diagnostics.F90 too. It does not require to look deeper than on the surface and just two 1 minute runs.

Let's just add the following PRINT* line to this subroutine before or after these lines (line around 393), it does not matter,

       CALL sdf_open(sdf_handle, full_filename, comm, c_sdf_write)>

The line we will add is telling us what indices of 3D e-field array each core is dealing with when they go to be saved:

print*, 'cell_x_min(rank+1),cell_x_max(rank+1), cell_y_min(rank+1),cell_y_max(rank+1), cell_z_min(rank+1),cell_z_max(rank+1)=', &  cell_x_min(rank+1),cell_x_max(rank+1), cell_y_min(rank+1),cell_y_max(rank+1), cell_z_min(rank+1),cell_z_max(rank+1) 

and let's run the standard 3D demo deck called cone.deck from epoch3d samples. With 1 core we get of course the entire array and initial and final indices for x,y and z dimensions look like this: 1 250 1 250 1 250 because nx, ny and nz all are =250 in initial settings.

Now let's run with 2 cores. Natural to expect that the entire domain will be divided by two equal parts. What we get for first core
cell_x_min(rank+1),cell_x_max(rank+1), cell_y_min(rank+1),cell_y_max(rank+1), cell_z_min(rank+1),cell_z_max(rank+1)=           1         250           1         250           1         125
is exactly what we expected, the core 0 took exactly half of the cube. But what we get for another core looks shocking:

 cell_x_min(rank+1),cell_x_max(rank+1), cell_y_min(rank+1),cell_y_max(rank+1), cell_z_min(rank+1),cell_z_max(rank+1)=       22050           5       32739       32739         126         250

What is that???
Z array is right by the way, it took exactly another half of the array from 126 to 250.

**Comment by Status-Mirror** [comment]
Hey Dan,

Let's consider the `cell_x_max` array. By navigating to `epoch3d/src` and running the Linux terminal command: `grep -rnw '.' -e 'cell_x_max'`, we are told all the files this array appears in.

We can see that it is initially allocated in `mpi_routines.F90`, and it is allocated to a size of `nprocx`. Here, `nprocx` is the number of processors in the `x` direction, which in your example is 1 (split is in $z$, not $x$). Hence, when you call `cell_x_max(rank+1)`, you are calling a number outside the bounds of the array, which explains the nonsense result.

To further verify this, if you ran `print*, cell_x_max`, you would only print a single number.

By sheer coincidence, we also have `nprocz=2`. Because there are no splits in $x$ and $y$, your `rank` output will actually work here - although this is misleading, as it will only "work" when your domain is split in 1 dimension. 

Hope this helps,
Stuart

**Comment by DanRRRR** [comment]
Wow! If these cell_x_min, cell_x_max, cell_y_min, cell_y_max, cell_z_min and cell_z_max not defined for all cores even if there was no division in appropriate dimension that makes the logic to find indices of decomposed for each core grid pretty complex.

 If i want to do the independent text/binary output of the e-field,  how to find them for the each sub-array (what each core is dumping in this subroutine) ?

**Comment by Status-Mirror** [comment]
There may be some misunderstanding - all ranks can see the locations of all splits. These arrays don't refer to the cell boundaries of the current cell.

- The value of `cell_x_max` on ranks 0 and 1 are both (250).
- The value of `cell_z_max` on ranks 0 and 1 are both (125, 250).

The entire field does _not_ exist in memory on any one MPI-rank. This is why each processor writes a section of the global array to a pre-determined point of memory in the output file, using the SDF library. For the fields, each rank prints: `ex(1:nx,1:ny,1:nz)`, where `nx`, `ny` and `nz` are related to the local size of the rank, and are different on each rank.

**Comment by DanRRRR** [comment]
One of developers wrote me this: "the decomposition of the grid is stored in the variables cell_x_min, cell_x_max, cell_y_min, cell_y_max, cell_z_min and cell_z_max. Each of these are 1D integer arrays that contain the maximum and minimum values in the hypothetical global array for the part of the fields that each processor holds. So if you want to find the minimum index in X for the current rank then that would be "cell_x_min(rank)"."

Was he wrong?

**Comment by Status-Mirror** [comment]
I agree with most of what the the last developer said, but I think a mistake was made in how to access the information from `cell_x_min`. I believe both `cell_x_min(rank)` and `cell_x_min(rank+1)` are wrong.

In 3D, domains are split along 3 directions: $(x, y, z)$. In your example, the domain decomposition is $(1\times 1 \times 2)$. It's not the _rank_ that's important, it's the position of the domain on the decomposition grid. This isn't found using `rank`, but instead using `x_coords`, `y_coords` and `z_coords`.

I think you should be using `cell_x_min(x_coords+1)`, `cell_y_min(y_coords+1)`, `cell_z_min(z_coords+1)`.

It's worth noting I didn't develop the core MPI and output routines, so I only have access to the same information as you. But, from what I can see, I think this should work for you.

Cheers,
Stuart

**Comment by DanRRRR** [comment]
Thanks for the hint to explore, Stuart!

**Comment by DanRRRR** [comment]
I was looking at the email notifications and in this GitHub few times per day and when did that today found that your respond was posted 4 days ago. I think i need to employ ChatGPT as a secretary to control numerous things around me, as i am totally out of any schedules lately. :)))

 Thanks for the hint and the demo code to try.  Hope the way i did also will work after substituting Ex, Ey and Ez i used to array as the array is the local entity to storing field for each core while Ex, Ey and Ez are global ones and this is why my approach did not work.

As to SDF as you remember i tried to deal with it after my failed attempt  
1) to make anyone who developed SDF to cooperate with Visit and Paraview speifically for Windows to adopt it for their visualizers. Visit and Paraview expressed their willingness but no one responded from SDF side. Which means the SDF is dead
2) to convince anyone to develop standalone converter and structure visualizer of SDF same way as there exist converters and visualizers for HDF files. With such tools anyone would extract any data from black box which could in long run  not supported anymore. We have to keep published data for 10 years, this is what top journals require. Can we keep data in dead format which seems is dead long ago?
3) as time progressed the software community finally essentially solved the problem of compatibility of binary formats and hence there is no need in any formats anymore. 200 different formats it supports caused Visit to compile for the entire day on a few dozen of cores and at the end to fail which happens on not that new anymore three years old OS like Ubuntu 22.
4) Fortran STREAM binary format is way faster and more compact than any proprietary formats and is is damage resistant, it is compatible to any other hardware or compiler now so why bother with formats which have no future ?
5) Also, EPOCH has the ability to read users external initial binary or unformatted file data but why it has no ability to dump its generated data in binary form?  Users will be able to use this data with absolutely any compilers or other software  they are familiar with. Instead we have only IDL, Matlab and Python as choices

**Comment by DanRRRR** [comment]
I finally succeeded on accessing field data. Thanks again, without this help if would take forever.
Please do something with SDF. 

SDF quietly killing EPOCH reducing its adoption several times. 
Please add direct output of all data. Or at least keeping SDF add HDF5 at least which is much more widely adopted and has many tools to extract the data
